### PR TITLE
fix: revoke OpenShift OAuth token server-side and prevent re-login loop on logout

### LINKED ISSUE #708: Logout does not work with OpenShift OAuth provider — user is immediately re-authenticated
## Description

Clicking \"Logout\" in the FlightCtl UI has no visible effect when using the OpenShift OAuth provider. The session cookie is cleared, but the user is silently re-authenticated within seconds and lands back on the dashboard as if nothing happened.

## Steps to reproduce

1. Deploy the FlightCtl UI with an OpenShift OAuth provider configured as the only (or default) auth provider.
2. Log in as cluster admin.
3. Open the user menu (top-right) and click **Logout**.

## Expected behavior

The user is logged out and must enter credentials again to regain access.

## Actual behavior

The page briefly shows a loading spinner, then the user is returned to the dashboard, still logged in.

## Root cause

PR #393 (EDM-2612, December 2025) removed OpenShift OAuth session termination from `OpenShiftAuthHandler.Logout`. The original code called `{apiServerURL}/.well-known/oauth-authorization-server` to discover the issuer and redirected to `{issuer}/logout`. This was removed because when `apiServerURL` was the K8s API server (`https://api.cluster:6443`), the discovery returned that same URL as the issuer, resulting in a redirect to `https://api.cluster:6443/logout` — which is not a valid endpoint.

Rather than fixing the URL derivation, the removal eliminated OpenShift session termination entirely. As a result:

1. FlightCtl session cookie is cleared ✓
2. Page reloads → 401 → redirects to `/login`
3. Login page detects a single non-K8s provider and **immediately** re-initiates the OAuth flow
4. OpenShift OAuth session is still active → new auth code issued silently
5. New token is stored → user is back on the dashboard

## Fix

Derive the logout URL directly from `authURL` (the OAuth authorization endpoint), which always points to the OAuth server — not the K8s API server. Taking `scheme://host` from `authURL` and appending `/logout` gives the correct `https://oauth-openshift.apps.{cluster}/logout` in all valid configurations, without any HTTP discovery round-trip.

A fix with regression tests is available: [link to PR].

## Environment

- **Cluster type:** Single Node OpenShift (SNO), ephemeral
- **FlightCtl UI version:** `66e5e623` (Helm chart installed from `main` at that commit)
- **Auth provider:** OpenShift OAuth, configured as the only provider
- **User:** cluster-admin

### PR DESCRIPTION
**Pull request by jkilzi**
Fixes #708

## Summary

- `OpenShiftAuthHandler.Logout` was a no-op since PR #393 (EDM-2612), clearing only the FlightCtl session cookie but leaving the OpenShift OAuth session alive
- With a single non-K8s provider configured, `LoginPage` immediately re-initiated the OAuth flow → users were silently re-authenticated after clicking Logout
- Redirecting the browser to `{oauth-server}/logout` (the approach attempted in the previous revision of this PR) returned HTTP **405 Method Not Allowed** for unauthenticated GET requests

### Fix — proxy (Go)

`OpenShiftAuthHandler.Logout` now revokes the access token server-side:

```
DELETE /apis/oauth.openshift.io/v1/oauthaccesstokens/{name}
Authorization: Bearer <token>
```

The token resource name is derived from the raw token value: `sha256~`-prefixed tokens are hashed with SHA-256 and base64url-encoded; legacy tokens are used as-is. The proxy always returns an empty redirect URL — no browser redirect needed.

### Fix — frontend (TypeScript)

To prevent the silent re-login loop after the session is cleared:

- `logout()` in `apiCalls.ts` sets a `sessionStorage` flag (`flightctl.justLoggedOut`) before navigating to `/login`
- `LoginPage` reads and clears the flag; when set, it skips the single-provider auto-redirect (`suppressAutoSelect`) and displays the provider selector so the user must click explicitly to log back in

## Test plan

- [x] `TestOpenShiftTokenName_SHA256Prefix` — correct SHA-256 + base64url derivation for `sha256~` tokens
- [x] `TestOpenShiftTokenName_LegacyToken` — legacy token returned unchanged
- [x] `TestOpenShiftLogout_RevokesTokenViaDelete` — DELETE is issued to the correct URL with Bearer auth; 200/204 treated as success
- [x] `TestOpenShift_LogoutRevocationLogged_OnUnexpectedStatus` — unexpected status is logged as a warning, not an error
- [x] `TestOpenShiftLogout_NoOpWhenTokenEmpty` — no HTTP call when token is empty
- [x] `TestOpenShiftLogout_NoOpWhenAPIServerURLEmpty` — no HTTP call when `apiServerURL` is unset
- [x] Manual: click Logout on a cluster using the OpenShift OAuth provider; browser lands on `/login` showing the provider selector without auto-redirecting

### CODE REVIEW

File: proxy/auth/helpers_test.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+func redirectBaseMatchesRequest(r *http.Request, u *url.URL) error {
+	ok, err := isSameSchemeAndHost(u.String(), r)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		return fmt.Errorf("origin mismatch: redirect base %q does not match request origin", u)
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (jkilzi) [reply proxy/auth/helpers_test.go:30]: Fixed in 4eec374. Added `*testing.T` as the first parameter and `t.Helper()` at the top of the function body. Also updated all three existing call sites in `redirect_test.go` to pass `t`.

File: proxy/auth/openshift_test.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	}
+
+	logoutURL, err := handler.Logout("test-token", "https://ui.example.com")
+	if err != nil {
+		t.Fatalf("expected no error on invalid authURL, got: %v", err)
+	}
+	if logoutURL != "" {
+		t.Fatalf("expected empty logout URL for invalid authURL, got: %q", logoutURL)
+	}
+}
```
[the remark this answers is not in the document]
Comment (jkilzi) [reply proxy/auth/openshift_test.go]: Fixed in 4eec374. Consolidated the four individual test functions into a single table-driven `TestOpenShiftLogout` with subtests.

File: proxy/auth/openshift.go [hunk 3]
```diff
@@ -123,9 +123,26 @@ func (o *OpenShiftAuthHandler) openshiftClientForRedirect(redirectURI string) (*
...
+	// EDM-2612 because it produced a K8s API server URL as the logout target
+	// (e.g. https://api.cluster:6443/logout) which is not a valid endpoint.
+	if o.authURL == "" {
+		return "", nil
+	}
+	parsed, err := url.Parse(o.authURL)
+	if err != nil || parsed.Host == "" {
+		return "", nil
+	}
+	return parsed.Scheme + "://" + parsed.Host + "/logout", nil
```
[the remark this answers is not in the document]
Comment (jkilzi) [reply proxy/auth/openshift.go]: Fixed in 4eec374. Added `parsed.Scheme == ""` to the guard and switched to `url.URL{Scheme: ..., Host: ..., Path: "/logout"}.String()` for URL construction.

**Comment by jkilzi** [comment]
Regarding the **Unchecked-Errors** warning: the `url.Parse` error on line 141 is checked — when it returns a non-nil error the function falls back to `("", nil)`. This is intentional: a logout URL derivation failure should degrade gracefully (no logout redirect, session cookie is still cleared) rather than surface an error to the caller, which would block the logout flow. There is nothing actionable the caller can do with a parse error here, so returning it would only add noise.

File: proxy/auth/openshift_test.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+		{
+			name:    "empty auth URL falls back",
+			authURL: "",
+			wantURL: "",
+		},
+		{
+			name:    "invalid auth URL falls back",
+			authURL: "://not-a-valid-url",
+			wantURL: "",
+		},
```
[the remark this answers is not in the document]
Comment (jkilzi) [reply proxy/auth/openshift_test.go:150]: The scheme-less URL concern has been superseded. The intermediate implementation that constructed a logout redirect URL from `authURL` (which is where the empty-Scheme risk lived) was replaced entirely in commit 6257b75c with server-side OAuth token revocation. The current `Logout()` method issues a `DELETE /apis/oauth.openshift.io/v1/oauthaccesstokens/{name}` and always returns `("", nil)` — it never builds a URL from `authURL`, so there is no code path that could produce a malformed `://…` redirect. No test case is needed for a code path that no longer exists.

File: proxy/auth/openshift.go [hunk 5]
```diff
@@ -123,11 +131,63 @@ func (o *OpenShiftAuthHandler) openshiftClientForRedirect(redirectURI string) (*
...
+// lands the user on the login page with no valid session.
 func (o *OpenShiftAuthHandler) Logout(token string, _ string) (string, error) {
-	// The cookie will be cleared by the proxy
+	if token == "" || o.apiServerURL == "" {
+		return "", nil
+	}
+
+	tokenName := openShiftTokenName(token)
+	revokeURL := fmt.Sprintf("%s/apis/oauth.openshift.io/v1/oauthaccesstokens/%s",
+		strings.TrimSuffix(o.apiServerURL, "/"), tokenName)
```
[the remark this answers is not in the document]
Comment (jkilzi) [reply proxy/auth/openshift.go]: Fixed in 19af6f4c. Added `openShiftAPIServerBase()` which parses the raw URL and returns only `scheme://host`, stripping any path, query, and fragment. The revocation URL is now built from that normalised base, so a URL like `https://oauth.example.com/oauth/authorize?foo=bar` correctly yields `https://oauth.example.com/apis/oauth.openshift.io/v1/oauthaccesstokens/{name}`. A dedicated `TestOpenShiftAPIServerBase` table test covers all edge cases, and `TestOpenShiftLogout` gains a case where `apiServerURL` carries a path+query to verify end-to-end normalization.

File: proxy/auth/openshift.go [hunk 6]
```diff
@@ -123,11 +131,63 @@ func (o *OpenShiftAuthHandler) openshiftClientForRedirect(redirectURI string) (*
...
+
+	req, err := http.NewRequest(http.MethodDelete, revokeURL, nil)
+	if err != nil {
+		log.GetLogger().WithError(err).Warn("Failed to build OpenShift token revocation request")
+		return "", nil
+	}
+	req.Header.Set("Authorization", "Bearer "+token)
+
+	resp, err := client.Do(req)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (jkilzi) [reply proxy/auth/openshift.go:190]: Fixed in 19af6f4c. Added `Timeout: 10 * time.Second` to the `http.Client` used for token revocation. If the API server stalls, the logout call now fails fast with an error after 10 seconds rather than blocking the proxy handler indefinitely.

**Comment by celdrake** [comment]
Thanks! I see for OCP (kubeadmin) it's behaving better now because it brings the user to the selection screen.
I will check this for other deployment methods and auth providers in a few weeks (after 1.3 code freeze).
